20240129-satellite-managed-encryption.md

```diff
@@ -0,0 +1,164 @@
+However, there are [security considerations](https://github.com/GoogleCloudPlatform/secrets-store-csi-driver-provider-gcp?tab=readme-ov-file#security-considerations) we have to make.
+
+#### Satellite Project Updates
+
+A new column should be added to the `projects` table to be the passphrase store, like `projects.passphrase_enc`. This should be the encrypted passphrase for satellite-managed-encryption projects only and `null` for others.
+Projects that are created before this feature is deployed and projects created after but opt out should have `null` in this column.
+
+Functionality should then be added to allow creating a project with a satellite managed encryption passphrase:
+* the KMS service generates a cryptographically random passphrase and encrypts it using the master key
+* the passphrase store is updated with the encrypted passphrase for this project
+* endpoints on the satellite API are added or updated so that an authenticated user can request the encryption passphrase for a project during access creation
+  * endpoints handlers should use the KMS service to retrieve the encrypted passphrase from the store and decrypt it using the master key
+
+#### Satellite UI Updates
+
+When creating a project, the user should be provided the option to create different types of projects:
+
+* satellite-managed-encryption project
+* user-managed-encryption project
+
+The user-managed-encryption project will be the same as before, and user will always be prompted to enter encryption passphrase when necessary, e.g. when using the file browser or generating an access grant.
+
+The satellite-managed-encryption project will never prompt the user for an encryption passphrase in the UI. Instead, when the user does an action that requires generating credentials, the UI will retrieve the project's encryption passphrase from the satellite KMS, and generate an access grant using this passphrase, and path encryption disabled.
+
+The user should not be able to directly retrieve a "raw API key" from the Satellite UI. This prevents the user from unintentionally creating "invalid" credentials using a raw API key with Uplink. The only credentials that can be generated from the access page for a satellite-managed-encryption project should be credentials that already have embedded encryption rules.
+
+The file browser in the UI should generate access credentials using the passphrase from the satellite KMS, and no path encryption.
+
+#### External Updates
+
+###### WebAssembly
+
+* Wasm functionality must be updated to allow the Satellite UI to properly generate access credentials with path encryption disabled. [Code for reference](https://github.com/storj/storj/blob/a62929fd5757a40eda5d0b044ad2cefc14708410/satellite/console/consolewasm/access.go#L22-L29).
+
+###### Uplink and Bindings
```

Comment:
We should make it clear that this section is not 100% necessary. We can choose not to update Uplink and bindings, and as long as the UI is implemented correctly, nothing should break.

I think either:
* this heading should be prominently marked as "optional", or
* this section should be moved out of the main design section, and into a "possible future improvements" section (or even to "out of scope" at the bottom)

Reply:
the end of the paragraph says _These changes are technically unnecessary if the UI is updated to not provide raw API keys for satellite-managed-encryption projects_. Should more emphasis be made?

Reply:
I guess that's fine - as long as it's mentioned. I think it would help to at least emphasize it (e.g. bold). I guess I just want it to be very obvious to anyone reading this document that we may not make _any_ changes to Uplink and bindings as part of this design.